File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.io.*;
import java.util.*;
public class Main {
    static int N, C;
    static int[] house;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());

        house = new int[N];

        for(int i=0;i<N;i++) {
            house[i] = Integer.parseInt(br.readLine());
        }
        Arrays.sort(house);

        int lo = 0;
        long hi = house[N-1]-house[0];
        int ans=0;

        while(lo <= hi) {
            int mid = (int) ((lo + hi) / 2);

            if(canInstall(mid)) {
                ans = mid;
                lo = mid + 1;
            } else {
                hi = mid-1;
            }
        }

        System.out.println(ans);
    }

    private static boolean canInstall(int dist) {
        int count = 1;
        int last = house[0];
        for(int i=0;i<N;i++) {
            if(house[i] - last >= dist) {
                count++;
                last=house[i];
            }
        }
        return count >= C;
    }
}

File: 백준/Gold/2110. 공유기 설치/공유기 설치.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

static int N, C;
static vector<int> house;

static bool canInstall(long long dist)
{
    int count = 1;
    int last = house[0];

    for (int i = 0; i < N; i++)
    {
        long long diff = static_cast<long long>(house[i]) - last;
        if (diff >= dist)
        {
            count++;
            last = house[i];
        }
    }

    return count >= C;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> C;
    house.resize(N);
    for (int i = 0; i < N; i++)
    {
        cin >> house[i];
    }
    sort(house.begin(), house.end());

    long long lo = 1;
    long long hi = static_cast<long long>(house[N - 1] - house[0]);
    int ans = 0;
    while (lo <= hi)
    {
        long long midLL = (lo + hi) / 2; // (lo+hi) / 2
        int mid = static_cast<int>(midLL);

        if (canInstall(mid))
        {
            ans = mid;
            lo = mid + 1;
        }
        else
        {
            hi = mid - 1;
        }
    }

    cout << ans;
}